File: horizons_py/horizons/memory.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .client import HorizonsClient
from . import models


class MemoryAPI:
    def __init__(self, client: HorizonsClient) -> None:
        self._client = client

    async def retrieve(self, *, agent_id: str, q: str, limit: int = 20) -> List[models.MemoryItem]:
        params = {"agent_id": agent_id, "q": q, "limit": limit}
        resp = await self._client._request("GET", "/api/v1/memory", params=params)
        data = await self._client.json(resp)
        return [models.MemoryItem.model_validate(item) for item in data]

    async def append(
        self,
        *,
        agent_id: str,
        item_type: models.MemoryType,
        content: Any,
        index_text: Optional[str] = None,
        importance_0_to_1: Optional[float] = None,
        created_at: Optional[str] = None,
    ) -> str:
        body: Dict[str, Any] = {
            "agent_id": agent_id,
            "item_type": item_type.value,
            "content": content,
        }
        if index_text:
            body["index_text"] = index_text
        if importance_0_to_1 is not None:
            body["importance_0_to_1"] = importance_0_to_1
        if created_at:
            body["created_at"] = created_at
        resp = await self._client._request("POST", "/api/v1/memory", json=body)
        data = await self._client.json(resp)
        return data["id"]

    async def summarize(self, *, agent_id: str, horizon: str) -> models.Summary:
        body = {"agent_id": agent_id, "horizon": horizon}
        resp = await self._client._request("POST", "/api/v1/memory/summarize", json=body)
        data = await self._client.json(resp)
        return models.Summary.model_validate(data)

File: horizons_ts/src/memory.ts
import { HorizonsClient } from "./client";
import { UUID } from "./types";

export type MemoryType = "observation" | "summary" | "action";

export interface MemoryItem {
  id?: string;
  scope: Record<string, string>;
  item_type: MemoryType;
  content: unknown;
  index_text?: string;
  importance_0_to_1?: number;
  created_at: string;
}

export interface Summary {
  agent_id: string;
  horizon: string;
  content: unknown;
  at: string;
}

export class MemoryAPI {
  constructor(private client: HorizonsClient) {}

  async retrieve(agent_id: string, q: string, limit = 20): Promise<MemoryItem[]> {
    const qs = new URLSearchParams({ agent_id, q, limit: String(limit) });
    return await this.client.request<MemoryItem[]>(`/api/v1/memory?${qs.toString()}`, { method: "GET" });
  }

  async append(input: {
    agent_id: string;
    item_type: MemoryType;
    content: unknown;
    index_text?: string;
    importance_0_to_1?: number;
    created_at?: string;
  }): Promise<string> {
    const resp = await this.client.request<{ id: string }>("/api/v1/memory", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(input),
    });
    return resp.id;
  }

  async summarize(agent_id: string, horizon: string): Promise<Summary> {
    const resp = await this.client.request<Summary>("/api/v1/memory/summarize", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ agent_id, horizon }),
    });
    return resp;
  }
}
